src/ui/ui_settings/ui_settings.js

```diff
@@ -0,0 +1,119 @@
+import { defaultsDeep } from 'lodash';
+import Bluebird from 'bluebird';
+
+import { getDefaultSettings } from './defaults';
+
+function hydrateUserSettings(user) {
+  return Object.keys(user)
+    .map(key => ({ key, userValue: user[key] }))
+    .filter(({ userValue }) => userValue !== null)
+    .reduce((acc, { key, userValue }) => ({ ...acc, [key]: { userValue } }), {});
+}
+
+function assertRequest(req) {
+  if (
```

Comment:
__issue:__ sorry, my bad, haven't noticed that earlier: it looks like this condition should have `||` all over the place, with `&&` it does the wrong thing (e.g. it won't throw if `req === {}`).

And that means we need a unit test that would have caught that :)

__bikeshedding:__ maybe guard against `req.headers === null` too (just for consistency's sake, but it's definitely up to you).